src/Compilers/Core/Portable/DiagnosticAnalyzer/ArtifactProducer.cs

```diff
@@ -0,0 +1,115 @@
+#else
+                using var writer = new StreamWriter(stream, Encoding.UTF8, bufferSize: DefaultBufferSize, leaveOpen: true);
+#endif
+                writer.Write(source);
+            });
+        }
+
+        /// <summary>
+        /// Generates an artifact with the contents of <paramref name="builder"/>.  The artifact will be written out with utf8 encoding.
+        /// </summary>
+        /// <param name="fileName">The file name to generate this artifact into.  Will be concatenated with the
+        /// generatedartifactsout path provided to the compiler.</param>
+        /// <param name="builder">The string builder containing the contents to generate</param>
+        protected void WriteArtifact(string fileName, StringBuilder builder)
+        {
+            WriteArtifact(fileName, stream =>
+            {
+#if NETCOREAPP
+                using var writer = new StreamWriter(stream, Encoding.UTF8, leaveOpen: true);
+#else
+                using var writer = new StreamWriter(stream, Encoding.UTF8, bufferSize: DefaultBufferSize, leaveOpen: true);
+#endif
+                writer.Write(builder);
+            });
+        }
+
+        /// <summary>
+        /// Generates an artifact with the contents and encoding specified by <paramref name="sourceText"/>.
+        /// </summary>
+        /// <param name="fileName">The file name to generate this artifact into.  Will be concatenated with the
+        /// generatedartifactsout path provided to the compiler.</param>
+        /// <param name="sourceText">The <see cref="SourceText"/> to generate</param>
+        protected void WriteArtifact(string fileName, SourceText sourceText)
+        {
+            WriteArtifact(fileName, stream =>
+            {
+#if NETCOREAPP
+                using var writer = new StreamWriter(stream, sourceText.Encoding ?? Encoding.UTF8, leaveOpen: true);
+#else
+                using var writer = new StreamWriter(stream, sourceText.Encoding ?? Encoding.UTF8, bufferSize: DefaultBufferSize, leaveOpen: true);
+#endif
+                sourceText.Write(writer);
+            });
+        }
+
+        /// <summary>
+        /// Requests a fresh stream associated with the given <paramref name="fileName"/> to write artifact data into.
+        /// The callback does should not call <see cref="Stream.Close"/>, <see cref="Stream.Flush"/>, or <see
+        /// cref="Stream.Dispose()"/>.  This overload is useful if there is a large amount of data to write, or if there
+        /// is binary data to write.
+        /// </summary>
+        /// <param name="fileName">The file name to generate this artifact into.  Will be concatenated with the
+        /// generatedartifactsout path provided to the compiler.</param>
+        /// <param name="writeStream">A callback that will be passed the stream to write into.</param>
+        protected void WriteArtifact(string fileName, Action<Stream> writeStream)
+            => CreateArtifactStream!(fileName, writeStream);
```

Comment:
Could take the encoding in here, create the writer and have an `Action<StreamWriter>` instead.